### PR TITLE
Issue 137

### PR DESCRIPTION
**Pull request by jacobdotcosta**
Implemented creating an hetzner server using ansible playbook.

### CODE REVIEW

**Comment by cmoulliard** [comment]
I think that there is an issue as I dont think that you changed 235 files. Can you check @jacobdotcosta ?

**Comment by jacobdotcosta** [comment]
> I think that there is an issue as I dont think that you changed 235 files. Can you check @jacobdotcosta ?

I moved the roles directory outside playbooks as it shouldn't be inside playbooks. That's why there are so many changes.

File: ansible/playbook/README.md [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+```
+
+## create hcloud server
+
+Parameters:
+
+| Variable | Default | Meaning |
+| --- | --- | --- |
+| vm_name |  | Name of the VM to be created at hetzner. |
+| server_type | cx41 | Server type. The list can be obtained using `hcloud server-type list` |
```
Comment (cmoulliard) [inline ansible/playbook/README.md]: Default should be `cx31`

File: ansible/playbook/README.md [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+$ ansible-playbook playbook/hetzner-create-hcloud-server.yml
+```
+
+## create hcloud server
+
+Parameters:
+
+| Variable | Default | Meaning |
+| --- | --- | --- |
+| vm_name |  | Name of the VM to be created at hetzner. |
```
Comment (cmoulliard) [inline ansible/playbook/README.md]: When the column default id empty, that means that no default exists within the role ? 
Can we also specify what are the mandatory parameters for the different vars ?

File: ansible/playbook/templates/user-data.j2 [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+yum_repos:
+  - epel:
+      baseurl: http://download.fedoraproject.org/pub/epel/7/$basearch
+      enabled: true
+      gpgcheck: false
+      name: Extra Packages for Centos 7 - $basearch
+
+packages:
+  - ansible
```
Comment (cmoulliard) [inline ansible/playbook/templates/user-data.j2]: I think that we can remove it excepted if it is needed when we execute the role remotely. WDYT ?

File: ansible/playbook/README.md (hunk 1 above)
Comment (jacobdotcosta) [reply ansible/playbook/README.md]: The default is something that the prompt assigns if we press intro without value. What I did was, if the variable is defined from the command line, Ansible doesn't prompt.

File: ansible/playbook/README.md (hunk 2 above)
Comment (jacobdotcosta) [reply ansible/playbook/README.md]: I have doubts regarding this column, The document might be better of without it. There are no mandatory parameters, one can prevent variables from being evaluated with if defined. IMO all prompts are required, optional variables aren't in prompts.

**Comment by cmoulliard** [comment]
We need certainly also a task or a job to shutdown the VM every 24 h. To be discussed with the team to see what best approach they prefer @jacobdotcosta

File: ansible/playbook/templates/user-data.j2 (hunk 3 above)
Comment (jacobdotcosta) [reply ansible/playbook/templates/user-data.j2]: I agree, we don't need ansible installed ootb.

File: ansible/playbook/README.md (hunk 1 above)
Comment (jacobdotcosta) [reply ansible/playbook/README.md]: What I can do is replace prompts with ansible tasks that assign the default value if it is not defined.

Comment (cmoulliard) [reply ansible/playbook/README.md]: As by example we will most of the time use as OS - `centos-7` and type - `cx31`, then defaults are needed. 

Is it possible to show a message when we use prompt to by example propose an example for the VM name (e.g ocp-311, k8s-115, ...) ? @jacobdotcosta

File: ansible/playbook/README.md (hunk 2 above)
Comment (cmoulliard) [reply ansible/playbook/README.md]: I will review the table after the merge and this is certainly something that we could do for the documentation in general

**Comment by cmoulliard** [comment]
LGTM. Do you still plan to add commits ? @jacobdotcosta

**Comment by jacobdotcosta** [comment]
> LGTM. Do you still plan to add commits ? @jacobdotcosta

I just did, improving documentation. With this one I think this is a good place to start.

File: ansible/playbook/README.md [hunk 4]
```diff
@@ -24,16 +24,22 @@ $ ansible-playbook playbook/hetzner-create-hcloud-server.yml
...
 ```bash
 $ ansible-playbook playbook/hetzner-create-hcloud-server.yml
-```
\ No newline at end of file
+```
+
+Optionally, using environment variables:
+
+```bash
+$ ansible-playbook playbook/hetzner-create-hcloud-server.yml -e '{ "vm_name": "my-name" }' -e '{ "salt_text": "<my salt>" }' -e '{ "password_text": "<my password>" }' -e '{ "hetzner_context_name": "<context name>" }' -e '{ "vm_delete": True }' -e '{ "pass_store_dir": "/home/<my home>/.my-password-store" }'
```
Comment (cmoulliard) [inline ansible/playbook/README.md]: Is there an alternative to `{}` as passing a parameter using JSON syntax is error prone and could be better handled using `-e key=value` ?

Comment (jacobdotcosta) [reply ansible/playbook/README.md]: Yes, we can use the form `-e key=value` but this has some issues that's why I purposed using json form. Booleans, for instance, are treated as String if we use `-e vm_delete=True` and it won't have the desired effect in the playbook. With booleans it's required to use the json form `-e '{ "vm_delete": True }'`.

File: ansible/playbook/hetzner-create-hcloud-server.yml [hunk 5]
```diff
@@ -51,20 +51,21 @@
...
-      when: "vm_delete is defined and vm_delete"
+      when: "vm_delete is defined and vm_delete | bool"
 
     - name: Delete cloud server key
       shell: "hcloud ssh-key delete snowdrop"
       register: hcloud_delete_key_res
       changed_when: "hcloud_delete_key_res.rc == 0"
-      when: "vm_delete is defined and vm_delete"
+      failed_when: "hcloud_delete_key_res.rc != 0 and ('SSH key not found' not in hcloud_delete_key_res.stderr)"
+      when: "vm_delete is defined and vm_delete | bool"
```
Comment (cmoulliard) [inline ansible/playbook/hetzner-create-hcloud-server.yml]: Will it be possible with this change to use then `-e vm_delete=true` ? @jacobdotcosta

Comment (jacobdotcosta) [reply ansible/playbook/hetzner-create-hcloud-server.yml]: Yes, I tried it and it works. 

```bash
$ ansible-playbook playbook/hetzner-create-hcloud-server.yml -e vm_name=myname -e salt_text=mysalt -e password_text=mypasswprd -e hetzner_context_name=mycontext -e vm_delete=true -e pass_store_dir="mypasswordstorelocation"
```

Comment (cmoulliard) [reply ansible/playbook/hetzner-create-hcloud-server.yml]: Perfect. I ill then merge the PR.

**Pull request merged**